wafer/kvpairs/models.py

```diff
@@ -0,0 +1,252 @@
+# -*- coding: utf-8 -*-
+# (c) 2015-16 martin f. krafft <madduck@debconf.org>
+# Released under the terms of the same licence as Wafer.
+
+from django.db import models
+from django.db.models import signals
+from django.core.exceptions import SuspiciousOperation, ValidationError
+from django.contrib.contenttypes.models import ContentType
+from django.utils.encoding import python_2_unicode_compatible
+
+from django.conf import settings
+KEYNAME_MAXLEN = settings.WAFER_KVPAIRS_KEYNAME_MAXLEN
+VALUE_MAXLEN = settings.WAFER_KVPAIRS_VALUE_MAXLEN
+AUTOCREATE_KEYS = settings.WAFER_KVPAIRS_AUTOCREATE_KEYS
+
+from .fields import IndirectGenericForeignKey
+
+
+# the list of models to which keys can be "attached", not sure why we limit
+# it, but it feels right. This feeds the choices parameter of the Key.model_ct
+# field, and can be disabled right here…
+SUPPORTED_MODEL_TYPES = ('sites.site', 'auth.user', 'users.userprofile',
+        'pages.page', 'schedule.venue', 'schedule.scheduleitem', 'talks.talk',
+        'sponsors.sponsorshippackages', 'sponsors.sponsor',
+        'tickets.tickettype', 'tickets.ticket')
+
+
+@python_2_unicode_compatible
+class Key(models.Model):
+    """Definition of a specific key related to another model"""
+
+    # Key ownership, which controls who can access/modify the values
+    owner = models.ForeignKey(settings.AUTH_USER_MODEL,
+            on_delete=models.PROTECT, # could be SET(get_sentinel_user)…
+            verbose_name='Key owner')
+    group = models.ForeignKey('auth.Group', null=True, # groups are not mandatory
+            on_delete=models.SET_NULL,
```

Comment:
I always try to avoid storing blank values that are not NULL. Form validation seems not to trip over this, at least not in the admin interface…